### ISSUE TITLE
no output from manysearch for failed aligments

### ISSUE THREAD

**Issue by peterjc** [sourmash-bio/sourmash_plugin_branchwater#542]
Example using two small FASTA files from https://github.com/pyani-plus/pyani-plus/tree/v0.0.1a1.dev2/tests/fixtures/bad_alignments which are intentionally not expected to have hashes in common:

* MGV-GENOME-0264574.fas
* MGV-GENOME-0357962.fna

Prepare signatures:

<details>

```console
❯ sourmash scripts singlesketch -I DNA -p "k=31,scaled=300" MGV-GENOME-0264574.fas -o MGV-GENOME-0264574.sig

== This is sourmash version 4.8.11. ==
== Please cite Irber et. al (2024), doi:10.21105/joss.06830. ==

=> sourmash_plugin_branchwater 0.9.11; cite Irber et al., doi: 10.1101/2022.11.02.514947

params: ['k=31,scaled=300,dna']
sketching file 'MGV-GENOME-0264574.fas' (DNA) with params 'k=31,scaled=300,dna' and name 'MGV-GENOME-0264574.fas' using a single thread
calculated 1 signatures for 1 sequences in MGV-GENOME-0264574.fas
...singlesketch is done! results in 'MGV-GENOME-0264574.sig'

❯ sourmash scripts singlesketch -I DNA -p "k=31,scaled=300" MGV-GENOME-0357962.fna -o MGV-GENOME-0357962.sig

== This is sourmash version 4.8.11. ==
== Please cite Irber et. al (2024), doi:10.21105/joss.06830. ==

=> sourmash_plugin_branchwater 0.9.11; cite Irber et al., doi: 10.1101/2022.11.02.514947

params: ['k=31,scaled=300,dna']
sketching file 'MGV-GENOME-0357962.fna' (DNA) with params 'k=31,scaled=300,dna' and name 'MGV-GENOME-0357962.fna' using a single thread
calculated 1 signatures for 1 sequences in MGV-GENOME-0357962.fna
...singlesketch is done! results in 'MGV-GENOME-0357962.sig'

❯ sourmash sig collect --quiet -F csv -o all_sigs.csv MGV-GENOME-*.sig
Loading signature information from MGV-GENOME-0264574.sig.
Loading signature information from MGV-GENOME-0357962.sig.
saved 2 manifest rows to 'all_sigs.csv'
```

</details>

Traditional ``sourmash compare`` outputs all four comparisons, with a value of zero for the off-diagonal bad comparisons (and prints a helpful warning).

```console
❯ sourmash compare --csv compare.csv --estimate-ani --containment -k 31 all_sigs.csv
== This is sourmash version 4.8.11. ==
== Please cite Irber et. al (2024), doi:10.21105/joss.06830. ==

loaded 2 signatures total.

WARNING: Some of these sketches may have no hashes in common based on chance alone (false negatives). Consider decreasing your scaled value to prevent this.
0-MGV-GENOME-0264...    [1. 0.]
1-MGV-GENOME-0357...    [0. 1.]
min similarity in matrix: 0.000

❯ cat compare.csv   
MGV-GENOME-0264574.fas,MGV-GENOME-0357962.fna
1.0,0.0
0.0,1.0
```

Using ``sourmash scripts manysearch`` we have partial output:

```console
❯ sourmash scripts manysearch -m DNA --quiet -o manysearch.csv all_sigs.csv all_sigs.csv
=> sourmash_plugin_branchwater 0.9.11; cite Irber et al., doi: 10.1101/2022.11.02.514947

ksize: 31 / scaled: None / moltype: DNA / threshold: 0.01
searching all sketches in 'all_sigs.csv' against 'all_sigs.csv' using 8 threads
selection scaled: None
Reading query(s) from: 'all_sigs.csv'
Loaded 2 query signature(s)
Setting scaled=300 based on max scaled in query collection
Reading search(s) from: 'all_sigs.csv'
Loaded 2 search signature(s)
DONE. Processed 2 search sigs

query             p_genome avg_abund   p_metag   metagenome name
--------          -------- ---------   -------   ---------------
MGV-GENOME-026457  100.0%     N/A        N/A     MGV-GENOME-026457
MGV-GENOME-035796  100.0%     N/A        N/A     MGV-GENOME-035796

❯ cat manysearch.csv                                                             
query_name,query_md5,match_name,containment,intersect_hashes,ksize,scaled,moltype,match_md5,jaccard,max_containment,query_containment_ani,match_containment_ani,average_containment_ani,max_containment_ani
MGV-GENOME-0357962.fna,94eb3e03e3df9b6276bf59c6ca2ec37e,MGV-GENOME-0357962.fna,1.0,270,31,300,DNA,94eb3e03e3df9b6276bf59c6ca2ec37e,1.0,1.0,1.0,1.0,1.0,1.0
MGV-GENOME-0264574.fas,8cf3615dcb9671c99f43e4afb7e0f9e8,MGV-GENOME-0264574.fas,1.0,111,31,300,DNA,8cf3615dcb9671c99f43e4afb7e0f9e8,1.0,1.0,1.0,1.0,1.0,1.0
```

Note there are only two entries in ``manysearch.csv`` rather than the four expected. The zero off-diagonal pairs are missing.

To be more explicit, here we request one of the bad pairings and get an empty output file:

```console
❯ sourmash scripts manysearch -m DNA --quiet -o bad-pair.csv MGV-GENOME-0357962.sig MGV-GENOME-0264574.sig
=> sourmash_plugin_branchwater 0.9.11; cite Irber et al., doi: 10.1101/2022.11.02.514947

ksize: 31 / scaled: None / moltype: DNA / threshold: 0.01
searching all sketches in 'MGV-GENOME-0357962.sig' against 'MGV-GENOME-0264574.sig' using 8 threads
selection scaled: None
Reading query(s) from: 'MGV-GENOME-0357962.sig'
Loaded 1 query signature(s)
Setting scaled=300 based on max scaled in query collection
Reading search(s) from: 'MGV-GENOME-0264574.sig'
Loaded 1 search signature(s)
DONE. Processed 1 search sigs

❯ wc bad-pair.csv  
0 0 0 bad-pair.csv
```

Expected behaviour: Output one line for *every* pairwise combination given, including bad alignments which might use zero or perhaps NA values.

**Comment by peterjc** [comment]
That makes sense, and would cover our use case nicely (where we want to partition a large all-vs-all comparison of many genomes).

I like ``--write-all`` but it would mean renaming the existing ``--write-all`` which is used already in ``pairwise``  for ``write self comparisons for all sketches`` to something like ``--write-self`` or ``--diagonal`` or ... - perhaps too late to change that.

I like ``--output-all-comparisons`` but it is long.

You asked for suggestions, so maybe ``--output-all`` or ``--output-fails`` for ``include failed comparisons in the output``, or ``--no-skip`` for ``don't skip failed comparisons`` (exact wording depending on how exactly things are implemented).

**Comment by ctb** [comment]
Am implementing this. One wrinkle: there's an explicit '-t/--threshold` option. When `--output-all-comparisons` is provided, what should we do with matches that are below that threshold, but non-zero?

I think I'll make it so that `--output-all-comparisons` explicitly ignores the threshold; that gels with the way the code is written, too.

**Comment by ctb** [comment]
(still curious why you're using `manysearch` and not `multisearch` BTW - `multisearch` should be much faster, with minimal memory consequence.)

**Comment by ctb** [comment]
Ooh, fun addendum: with `--output-all-comparisons`, manysearch against a RocksDB database will only report comparisons that have non-zero overlap, while manysearch against a collection of sketches will report all comparisons...

**Comment by ctb** [comment]
Adding `-A/--output-all-comparisons` in https://github.com/sourmash-bio/sourmash_plugin_branchwater/pull/544/files. @peterjc if you're interested, go on over and poke at it - lmk comments. Once it's merged I'll cut a new release.

**Comment by peterjc** [comment]
I'm not sure why I've been using ``manysearch`` over ``multisearch``, perhaps I was swayed by the help string suggesting the memory usage might be more of an issue than it is?

I should be able to try out the branch this week (assuming I don't get bogged down in building the rust code). Right now our branchwater manysearch wrapper is assuming missing comparisons failed to pass the thresholds and records them as null.

**Comment by ctb** [comment]
`-A/--output-all-comparisons` is now available in sourmash_plugin_branchwater v0.9.12. Enjoy!
